ArmVirtPkg/ArmVirtKvmTool.dsc

```diff
@@ -352,17 +347,14 @@
   ArmPkg/Drivers/ArmPciCpuIo2Dxe/ArmPciCpuIo2Dxe.inf {
     <LibraryClasses>
       NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
-      NULL|OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf

```

Comment:
Why is this removed?

Reply:
To fix CLANGDWARF build for ArmVirtKvmTool.dsc. 

ld.lld: error: duplicate symbol: PciExpressRegisterForRuntimeAccess
>>> defined in /home/mikhail/Amaranth/UDK/Build/ArmVirtKvmTool-AARCH64/DEBUG_CLANGDWARF/AARCH64/MdePkg/Library/BasePciExpressLib/BasePciExpressLib/OUTPUT/BasePciExpressLib.lib(PciExpressLib.obj)
>>> defined in /home/mikhail/Amaranth/UDK/Build/ArmVirtKvmTool-AARCH64/DEBUG_CLANGDWARF/AARCH64/OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib/OUTPUT/BaseCachingPciExpressLib.lib(PciExpressLib.obj)

...

Reply:
`PciExpressLib` is defined 2 times: 
1.  In `ArmVirtPkg/ArmVirt.dsc.inc` as  `PciExpressLib|OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf`
2.  In `ArmVirtPkg/ArmVirtKvmTool.dsc` as `PciExpressLib|MdePkg/Library/BasePciExpressLib/BasePciExpressLib.inf`

Removing definition in `ArmVirtPkg/ArmVirtKvmTool.dsc` leads to:
```
/home/mikhail/Amaranth/UDK/ArmVirtPkg/ArmVirtKvmTool.dsc(...): error 1001: Module type [UEFI_APPLICATION] is not supported by library instance [/home/mikhail/Amaranth/UDK/OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf]
	consumed by library instance [/home/mikhail/Amaranth/UDK/OvmfPkg/Library/BaseCachingPciExpressLib/BaseCachingPciExpressLib.inf] which is consumed by module [/home/mikhail/Amaranth/UDK/ShellPkg/Application/Shell/Shell.inf]
```

Reply:
Well, that’s not how DSC LibraryClasses work. Definitions of the same class *never* duplicate, they override each other by some odd rules (generally, the more specific scope takes precedence for when it applies).

You get the duplication because the override you removed is NULL, thus not tied to any specific LibraryClass and is always linked. Most likely those modules don’t have explicit dependencies on PciExpressLib, but some change in AUDK caused an implicit dependency, thus the default implementation clashes with this NULL one. That should probably investigated first, whether the new implicit dependency is legitimate.

if it is, it should be sufficient to change the LibraryClass name from NULL to PciExpressLib here. I don’t see a reason why you would force-link this library, as the constructor has no interesting side-effects.